File: DSA/Arshdeep-Kaur/Day 59/Day59_Question2.cpp
class Solution {
  public:
    Node* reverseList(Node* head) {
        Node* prev = NULL;
        Node* curr = head;
        Node* next = NULL;
    
        while (curr != NULL) {
            next = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = next;
        }
    
        return prev;
    }
    
    void insertNode(Node* &head, Node* &tail, int data) {
        //creating a new node
        Node* temp = new Node(data);
    
        //empty list
        if (head == NULL) {
            head = temp;
            tail = temp;
            return;
        }
        
        else {
            tail -> next = temp;
            tail = tail->next;
        }
    }
    
    Node* add(Node* first, Node* second) {
        int sum = 0;
        int carry = 0;
        int digit = 0;
        Node* ansHead = NULL;
        Node* ansTail = NULL;
    
        while (first != NULL || second != NULL || carry != 0) {
            int val1 = 0;
            if (first != NULL) {
                val1 = first -> data;
            }
    
            int val2 = 0;
            if (second != NULL) {
                val2 = second -> data;
            }
            
            sum = carry + val1 + val2;
            digit = sum % 10;
            insertNode(ansHead, ansTail, digit);
            carry = sum / 10;
    
            if (first != NULL) {
                first = first -> next;
            }
    
            if (second != NULL) {
                second = second -> next;
            }
        }
    
        return ansHead;
    }

    // Function to add two numbers represented by linked list.
    Node* addTwoLists(struct Node* num1, struct Node* num2) {
        //step1 - reverse input LL
        num1 = reverseList(num1);
        num2 = reverseList(num2);
    
        //step2 - add 2 LL
        Node* ans = add(num1, num2);
    
        //step3 - reverse the ans LL
        ans = reverseList(ans);
    
        return ans;
    }
};

File: DSA/Manav-Rajput/Day59_question2.java
class Solution {
    // Function to add two numbers represented by linked list.
    static Node addTwoLists(Node num1, Node num2) {
        // Reverse both lists to start addition from the least significant digit
        num1 = reverseList(num1);
        num2 = reverseList(num2);
        
        Node dummyHead = new Node(0);
        Node current = dummyHead;
        int carry = 0;
        
        // Traverse both lists
        while (num1 != null || num2 != null) {
            int sum = carry;
            
            if (num1 != null) {
                sum += num1.data;
                num1 = num1.next;
            }
            
            if (num2 != null) {
                sum += num2.data;
                num2 = num2.next;
            }
            
            // Calculate carry for the next iteration
            carry = sum / 10;
            sum = sum % 10;
            
            // Create a new node with the sum and attach to the result list
            current.next = new Node(sum);
            current = current.next;
        }
        
        // If there's any carry left, add it as a new node
        if (carry > 0) {
            current.next = new Node(carry);
        }
        
        // Reverse the resultant list to restore the proper order
        return reverseList(dummyHead.next);
    }
    
    // Helper method to reverse the linked list
    static Node reverseList(Node head) {
        Node prev = null;
        Node current = head;
        Node next = null;
        
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        
        return prev;
    }
}
